File: easy/Count Complete Tree Nodes.js
/**
 * @param {TreeNode} root
 * @return {number}
 */

/*
    思路 : 
        recursion traverse結束 , 我寫了遞迴,BFS/DFS , 前兩個效果差不多,DFS則是在空間上較優一點
*/

/*
    recursion
*/

var countNodes = function(root) {
    
    let result = 0 ;  

    let traverse = node => {
        if (node == null) {return} 
        else {
            result += 1 
            traverse(node.left) ; 
            traverse(node.right) ; 
            return 
        }

    };

    traverse(root) ; 

    return result ; 

};


/*
    BFS :  
*/

var countNodes = function(root) {

    if (root == null) {return 0 }
    let result = 0 ; 
    let queue = [] ; 
    
    queue.push(root) ; 

    while (queue.length >0) {

        let node = queue.shift() ; 
        result += 1 ; 
        if (node.left !=null) {queue.push(node.left)}
        if (node.right !=null) {queue.push(node.right)}

    }
    return result ; 
};

/*
    DFS :  
*/

var countNodes = function(root) {

    if (root == null) {return 0 }
    let result = 0 ; 
    let stack = [] ; 
    
    stack.push(root) ; 

    while (stack.length >0) {

        let node = stack.pop() ; 
        result += 1 ; 
        if (node.left !=null) {stack.push(node.left)}
        if (node.right !=null) {stack.push(node.right)}

    }
    return result ; 
};

File: RUSH/Count Complete Tree Nodes.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from typing import Optional

# 用低於O(N)的時間複雜度去計算complete tree的節點數量 , 
# complete binary tree除了最後一層以外每一層的節點數都等於 2^n , 
# 所以計算的方式,假設最後一層為n , 則總節點數 = 2 + 2^1 + 2^2 + ... 2^(n-1) + 最後一層節點數

class Solution:

    def countNodes(self, root: Optional[TreeNode]) -> int:
    

        def count(node) : 
            
            if node is None : return 0 
            
            return 1 + count(node.left) + count(node.right) 
        
        return count(root)
            